devops/PyPI-Release.yml

```diff
@@ -0,0 +1,200 @@
+
+- stage: TestWheel
+  displayName: Test Wheel from Artifact
+  dependsOn: CreateWheel
+  pool:
+    vmImage: $(poolImage)
+
+  jobs:
+    - template: templates/all-tests-job-template.yml
+      parameters:
+        platforms:  { Linux: ubuntu-latest, MacOS: macos-latest, Windows: windows-latest }
+        pyVersions: [3.6, 3.7]
+        installationType: 'WheelArtifact'
+        envArtifactStem: TestWheelFreeze
+        envFileStem: requirements-wheel-test
+        wheelArtifactName: $(packageArtifactName)
+
+# ==================================================================================================
+
+- stage: UploadWheel
+  displayName: Upload Wheel to PyPI (${{parameters.releaseType}})
+  dependsOn: TestWheel
+  pool:
+    vmImage: $(poolImage)
+
+  variables:
+    ${{ if eq(parameters.releaseType, 'Test')}}:
+      twineConnection: PyPI-Test
+      twineEndpoint: PyPITest
+    ${{ if eq(parameters.releaseType, 'Production')}}:
+      twineConnection: PyPI-Prod
+      twineEndpoint: PyPIProd
+
+  jobs:
+  - deployment: 'PyPI_${{parameters.releaseType}}_Upload'
+    displayName: PyPI ${{parameters.releaseType}} Upload
+    ${{ if eq(parameters.releaseType, 'Test')}}:
+      environment: 'PyPI-Test Deployment'
+    ${{ if eq(parameters.releaseType, 'Production')}}:
+      environment: 'PyPI Deployment'
+    pool:
+      vmImage: $(poolImage)
+
+    strategy:
+      runOnce:
+        deploy:
+          steps:
+          - task: UsePythonVersion@0
+            displayName: 'Use Python $(poolPythonVersion)'
+            inputs:
+              versionSpec: $(poolPythonVersion)
+              addToPath: true
+          
+          - script: pip install twine
+            displayName: 'Install twine'
+
+          - task: TwineAuthenticate@0
+            inputs:
+              externalFeeds: ${{variables.twineConnection}}
+
+          - script: 'twine upload --verbose -r $(twineEndpoint) --config-file $(PYPIRC_PATH) $(Pipeline.Workspace)/$(packageArtifactName)/*'
+            displayName: Upload to ${{parameters.releaseType}} PyPI
+
+        # TODO: Add GitHub Release task, so links in PyPI ReadMe will work without manual intervention (Prod only)
+  
+  - job: PyPI_Pause
+    pool: server
+    dependsOn: 'PyPI_${{parameters.releaseType}}_Upload'
+    displayName: PyPI Pause
+
+    steps:
+    - task: Delay@1
+      displayName: "Pause to allow PyPI updates to complete"
+      inputs:
+        delayForMinutes: "5"
+
+# # ==================================================================================================
+
+- stage: TestFromPyPI
+  displayName: Test package from ${{parameters.releaseType}} PyPI
+  dependsOn: UploadWheel
+  pool:
+    vmImage: $(poolImage)
+  
+  variables:
+    envInfoArtifact: TestPyPIFreeze
+    envInfoFileBase: requirements-pypi-test
+
+  jobs:
+  - template: templates/all-tests-job-template.yml
+    parameters:
+      platforms:  { Linux: ubuntu-latest, MacOS: macos-latest, Windows: windows-latest }
```

Comment:
the macos tests work?  I thought there were some tolerance issues?

Reply:
They appear to be intermittent, unfortunately. We might want to trim the testing when it comes time to do an actual release. I have managed to get all stages to work, but when tests are unreliable, we will have problems.